docs/architecture/adr-071-cryptography-v2.md

```diff
@@ -0,0 +1,566 @@
+- Create a new package `cryptoV2` and implement this ADR. Create unit tests, documentation and examples.
+- Deprecate the old crypto package. The old crypto package will still be usable, but it will be marked as deprecated and users can opt to use the new package.
+- Migrate the codebase to use the new `CryptoV2` package and remove the old crypto package.
+
+_A more detailed migration path is provided in the corresponding document._
+
+### Positive
+
+* Single place of truth
+* Easier to use interfaces
+* Easier to extend
+* Unit test for each crypto package
+* Greater maintainability
+* Incentivize addition of implementations instead of forks
+* Decoupling behaviour from implementation
+* Sanitization of code
+
+### Negative
+
+* It will involve an effort to adapt existing code.
+* It will require attention to detail and audition.
+
+### Neutral
+
+* It will involve extensive testing.
+
+## Test Cases
+
+*The code will be unit tested to ensure a high code coverage
+- There should be integration tests around Wallet, Keyring and CryptoProviders.
+- There should be benchmark tests for hashing, keyring, encryption, decryption, signing and verifying functions.
+
+## Further Discussions
+
+> While an ADR is in the DRAFT or PROPOSED stage, this section should contain a
+> summary of issues to be solved in future iterations (usually referencing comments
+> from a pull-request discussion).
+>
+> Later, this section can optionally list ideas or improvements the author or
+> reviewers found during the analysis of this ADR.
+
+
+## References
+
+* TPM 2.0 support: https://github.com/google/go-tpm
+* Initial basic PKCS#11
+  https://docs.oasis-open.org/pkcs11/pkcs11-base/v3.0/os/pkcs11-base-v3.0-os.pdf
+  https://docs.aws.amazon.com/cloudhsm/latest/userguide/pkcs11-library.html
+* https://docs.aws.amazon.com/cloudhsm/latest/userguide/pkcs11-key-types.html
+* https://solanacookbook.com/references/keypairs-and-wallets.html#how-to-generate-a-new-keypair
+* https://www.nist.gov/news-events/news/2022/07/nist-announces-first-four-quantum-resistant-cryptographic-algorithms
+* https://blog.cloudflare.com/nist-post-quantum-surprise/
+* https://pkg.go.dev/crypto#Hash
+* https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html
+
+## Appendix
+
+### Tentative Primitive Building Blocks
+
+This is a tentative list of primitives that we might want to support.
+This is not a final list or comprehensive, and it is subject to change.
+Moreover, it is important to emphasize the purpose of this work allows extensibility so any other primitive can be added in the future.
+
+* digital signatures
+    *  RSA (PSS)
+    *  ECDSA (secp256r1, secp256k1, etc.)
+    *  EdDSA (ed25519, ed448)
+    *  SR25519
+    *  Schnorr
+    *  Lattice-based (Dilithium)
+    *  BLS (BLS12-381, 377?)
+
+* encryption
+    * AES (AES-GCM, AES-CCM)
+    * RSA (OAEP)
+    * salsa20
+    * (x)chacha20 / (x)ChaCha20-Poly1305 (AEAD)
+    * Dilithium
```

Comment:
I assume this was meant to be Kyber, as the KEM based on some of the same primitives as Dilithium?

Reply:
(Note that I would recommend waiting for ML-KEM and ML-DSA to be standardized this summer.)

Reply:
Thanks for the recommendation! That's a tentative list of signatures that might be implemented after the basic architecture gets implemented. The idea is in the first major release to support only the curves and algos that are currently in the sdk's code + some implementation to serve as examples.
For sure this list will get refined, and most important, anyone will be able to add new signatures in a modular way.